pkg/controller/workload/bpfcache/locality_cache.go

```diff
@@ -0,0 +1,154 @@
+package bpfcache
+
+import (
+	"kmesh.net/kmesh/api/v2/workloadapi"
+)
+
+const (
+	REGION    = 1 << iota // 000001
+	ZONE                  // 000010
+	SUBZONE               // 000100
+	NODENAME              // 001000
+	CLUSTERID             // 010000
+	NETWORK               // 100000
+)
+
+type localityInfo struct {
+	region    string // init from workload.GetLocality().GetRegion()
+	zone      string // init from workload.GetLocality().GetZone()
+	subZone   string // init from workload.GetLocality().GetSubZone()
+	nodeName  string // init from os.Getenv("NODE_NAME"), workload.GetNode()
+	clusterId string // init from workload.GetClusterId()
+	network   string // workload.GetNetwork()
+	mask      uint32 // mask
+}
+
+func Valid(s string) bool {
+	return s != ""
+}
+
+func (l *localityInfo) Set(s string, param uint32) {
+	if !Valid(s) {
+		return
+	}
+	switch param {
+	case REGION:
+		l.region = s
+	case ZONE:
+		l.zone = s
+	case SUBZONE:
+		l.subZone = s
+	case NODENAME:
+		l.nodeName = s
+	case CLUSTERID:
+		l.clusterId = s
+	case NETWORK:
+		l.network = s
+	}
+	l.mask |= param
+}
+
+func (l *localityInfo) Clear(param uint32) {
+	l.mask &= ^param
+}
+
+func (l *localityInfo) IsSet(param uint32) bool {
+	return l.mask&param != 0
+}
+
+type LocalityCache struct {
+	LbPolicy               uint32
+	localityInfo           localityInfo
+	LbStrictIndex          uint32 // for failover strict mode
+	isLocalityInfoSet      bool
+	RoutingPreference      []workloadapi.LoadBalancing_Scope
+	isRoutingPreferenceSet bool
+	workloadWaitQueue      map[*workloadapi.Workload]struct{}
+}
+
+func NewLocalityCache() *LocalityCache {
+	return &LocalityCache{
+		localityInfo:           localityInfo{},
+		isLocalityInfoSet:      false,
+		RoutingPreference:      make([]workloadapi.LoadBalancing_Scope, 0),
+		isRoutingPreferenceSet: false,
+		workloadWaitQueue:      make(map[*workloadapi.Workload]struct{}),
+	}
+}
+
+func (l *LocalityCache) SetLocality(nodeName, clusterId, network string, locality *workloadapi.Locality) {
+	// notice: nodeName should set by processor or os.Getenv("NODE_NAME"),
+	l.localityInfo.Set(nodeName, NODENAME)
+	l.localityInfo.Set(locality.GetRegion(), REGION)
+	l.localityInfo.Set(locality.GetSubzone(), SUBZONE)
+	l.localityInfo.Set(locality.GetZone(), ZONE)
+	l.localityInfo.Set(clusterId, CLUSTERID)
+	l.localityInfo.Set(network, NETWORK)
+
+	l.isLocalityInfoSet = true
+}
+
+func (l *LocalityCache) SetRoutingPreference(s []workloadapi.LoadBalancing_Scope) {
+	// notice: s should set by lb.GetRoutingPreference()
+	if len(s) > 0 {
+		l.RoutingPreference = s
+		l.LbStrictIndex = uint32(len(s))
```

Comment:
Not sure i understand this, previously i thought this is a flag.

Reply:
In strict mode, only the workloads that exactly match every item in the routePreference are considered. In other words, only the workloads with a priority equal to len(routePreference) are taken into account.　We set LbStrictIndex so that the kernel BPF program filters out priorities other than lbStrictIndex in strict mode.